lib/pool.js

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * These options are passed into the [workerpool](https://npm.im/workerpool) module.
+ * @type {Partial<WorkerPoolOptions>}
+ */
+const WORKER_POOL_DEFAULT_OPTS = {
+  // use child processes, not worker threads!
+  workerType: 'process',
+  // ensure the same flags sent to `node` for this `mocha` invocation are passed
+  // along to children
+  forkOpts: {execArgv: process.execArgv},
+  maxWorkers: DEFAULT_MAX_WORKERS
+};
+
+/**
+ * A wrapper around a third-party worker pool implementation.
+ */
+class WorkerPool {
+  constructor(opts = WORKER_POOL_DEFAULT_OPTS) {
+    const maxWorkers = Math.max(1, opts.maxWorkers);
+
+    if (maxWorkers < 2) {
+      debug(
+        'not enough CPU cores available (%d) to run multiple jobs; avoid --parallel on this machine',
+        CPU_COUNT
+      );
+    } else if (maxWorkers >= CPU_COUNT) {
+      debug(
+        '%d concurrent job(s) requested, but only %d core(s) available',
+        maxWorkers,
+        CPU_COUNT
+      );
+    }
+    debug(
+      'run(): starting worker pool of max size %d, using node args: %s',
+      maxWorkers,
+      process.execArgv.join(' ')
+    );
+
+    this.options = Object.assign({}, opts, {maxWorkers});
+    this._pool = workerpool.pool(WORKER_PATH, this.options);
+  }
+
+  /**
+   * Terminates all workers in the pool.
+   * @param {boolean} [force] - Whether to force-kill workers. By default, lets workers finish their current task before termination.
+   * @private
+   * @returns {Promise<void>}
+   */
+  async terminate(force = false) {
+    return this._pool.terminate(force);
+  }
+
+  /**
+   * Adds a test file run to the worker pool queue for execution by a worker process.
+   *
+   * Handles serialization/deserialization.
+   *
+   * @param {string} filepath - Filepath of test
+   * @param {Options} [options] - Options for Mocha instance
+   * @private
+   * @returns {Promise<SerializedWorkerResult>}
+   */
+  async run(filepath, options = {}) {
+    if (!filepath || typeof filepath !== 'string') {
+      throw createInvalidArgumentTypeError(
+        'Expected a non-empty filepath',
+        'filepath',
+        'string'
+      );
+    }
+    const serializedOptions = WorkerPool.serializeOptions(options);
+    const result = await this._pool.exec('run', [filepath, serializedOptions]);
+    return deserialize(result);
+  }
+
+  /**
+   * Returns stats about the state of the worker processes in the pool.
+   *
+   * Used for debugging.
+   *
+   * @private
+   */
+  stats() {
+    return this._pool.stats();
+  }
+
+  /**
+   * Instantiates a {@link WorkerPool}.
+   */
+  static create(...args) {
+    return new WorkerPool(...args);
+  }
+
+  /**
+   * Given Mocha options object `opts`, serialize into a format suitable for
+   * transmission over IPC.
+   *
+   * @param {Options} [opts] - Mocha options
+   * @private
+   * @returns {string} Serialized options
+   */
+  static serializeOptions(opts = {}) {
+    if (!optionsCache.has(opts)) {
```

Comment:
hmm, I mean, practically speaking, yes, but I think we can do this more cleanly.  We need to bootstrap each worker process with options, and right now we're sending the options thru every time--we only really need the options once.  I didn't see an obvious way to call, e.g., `init()` on a worker and _then_ call `run()` on it without eagerly creating _max_workers_ workers.  Of course, we don't know which worker will receive the `run()` call, or if it's already been initialized.   This cache avoids having to re-serialize the options every time.  

So yeah this can likely be done a better way, but not sure if it's really worth the up-front investment.

Reply:
Ah I see what you mean. I think we need the `workerData` feature for that:

https://github.com/josdejong/workerpool/issues/65

So I guess it's fine if it lands like this and whenever issue 65 is implemented, refactor this piece of code.